### PR TITLE
Eliminate double layer hosting of video content when UI-Side compositing is enabled

### PR DESCRIPTION
**Pull request by jernoble**
#### 897c865996c5bcef29fb5b058df0dc6ae1172466
<pre>
Eliminate double layer hosting of video content when UI-Side compositing is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=252021">https://bugs.webkit.org/show_bug.cgi?id=252021</a>
&lt;rdar://105246033&gt;

Reviewed by Eric Carlson and Per Arne Vollan.

Currently, video content from the GPU process is remotely-hosted into the WebContent
process, and is subsequently remotely hosted from the WebContent process to the UI
process when UI-side compositing is enabled. This is unnecessary, and necessitates
having an active CARenderer in the Web Content process. Rather than double hosting
video content from GPU -&gt; WebContent -&gt; UI process, video content should be hosted
directly from the GPU process into the UI process.

To facilitate this, re-use the infrastructure built into VideoFullscreenManager to
handle the required XPC communication to send resize information back down from the
UI process to the GPU process.

HTMLMediaElement will need to expose API to resize its video layer, and to provide
the remote hosting context ID from the GPU process.

Rather than have a separate identifier of its own, VideoFullscreenManager and
PlaybackSessionManager can just use a HTMLMediaElement identifier for the purposes
of passing messages across the XPC boundary.

WebAVPlayerLayer must support use cases more like VideoLayerRemoteCocoa, including
handling multiple VideoGravity settings, so it is improved to take these settings
into account.

WebAVPlayerLayerView has been pulled out of VideoFullScreenManagerAVKit into its
own source and header files.

GraphicsLayerCA, RemoteLayerTreeHost, and RemoteLayerTree all must be modified to
accept a HTMLVideoElement (or its identifier) as a parameter.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ImageOverlay.cpp:
(WebCore::ImageOverlay::updateSubtree):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setVideoInlineSizeFenced):
(WebCore::HTMLMediaElement::layerHostingContextID):
(WebCore::HTMLMediaElement::naturalSize):
* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElement::identifier const):
* Source/WebCore/platform/cocoa/VideoFullscreenModel.h:
* Source/WebCore/platform/cocoa/VideoFullscreenModelVideoElement.h:
* Source/WebCore/platform/cocoa/VideoFullscreenModelVideoElement.mm:
(WebCore::VideoFullscreenModelVideoElement::setVideoInlineSizeFenced):
* Source/WebCore/platform/cocoa/WebAVPlayerLayer.h:
* Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm:
(-[WebAVPlayerLayer layoutSublayers]):
(-[WebAVPlayerLayer resolveBounds]):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::layerMode const):
(WebCore::GraphicsLayer::setContentsToVideoElement):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::setVideoInlineSizeFenced):
(WebCore::MediaPlayer::hostingContextID const):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::setVideoInlineSizeFenced):
(WebCore::MediaPlayerPrivateInterface::hostingContextID const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::createPlatformCALayer):
(WebCore::GraphicsLayerCA::setContentsToPlatformLayer):
(WebCore::GraphicsLayerCA::setContentsToVideoElement):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::PlatformCALayerCocoa):
* Source/WebCore/platform/graphics/ca/cocoa/WebVideoContainerLayer.mm:
(-[WebVideoContainerLayer setPosition:]): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.h:
* Source/WebCore/platform/mac/VideoFullscreenInterfaceMac.h:
* Source/WebCore/platform/mac/VideoFullscreenInterfaceMac.mm:
(-[WebVideoFullscreenInterfaceMacObjC setUpPIPForVideoView:withFrame:inWindow:]):
(WebCore::VideoFullscreenInterfaceMac::createLayer):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebKit/Platform/cocoa/LayerHostingContext.h:
* Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
(WebKit::LayerHostingContext::createForPort):
(WebKit::LayerHostingContext::createForExternalHostingProcess):
(WebKit::LayerHostingContext::createTransportLayerForRemoteHosting):
(WebKit::LayerHostingContext::updateCachedContextID):
(WebKit::LayerHostingContext::cachedContextID):
* Source/WebKit/Shared/PlaybackSessionContextIdentifier.h:
(): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::drawInContext):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::LayerCreationProperties):
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::encode const):
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::decode):
(WebKit::RemoteLayerTreeTransaction::description const):
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenManagerProxy::createLayerWithID):
(WebKit::VideoFullscreenManagerProxy::createLayerHostViewWithID):
(WebKit::VideoFullscreenManagerProxy::setupFullscreenWithID):
(WebKit::VideoFullscreenManagerProxy::didCleanupFullscreen):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::hostingContextID const):
(WebKit::MediaPlayerPrivateRemote::setLayerHostingContextID):
(WebKit::MediaPlayerPrivateRemote::setCachedPresentationSize):
(WebKit::MediaPlayerPrivateRemote::cachedPresentationSize):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/cocoa/MediaPlayerPrivateRemoteCocoa.mm:
(WebKit::MediaPlayerPrivateRemote::layerHostingContextIdChanged):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
(WebKit::GraphicsLayerCARemote::createPlatformCALayer):
(WebKit::GraphicsLayerCARemote::layerMode const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.cpp:
(WebKit::PlatformCALayerRemote::create):
(WebKit::PlatformCALayerRemote::recursiveBuildTransaction):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.mm:
(WebKit::PlatformCALayerRemoteCustom::create):
(WebKit::PlatformCALayerRemoteCustom::PlatformCALayerRemoteCustom):
(WebKit::PlatformCALayerRemoteCustom::hostingContextID):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.h:
(WebKit::RemoteLayerTreeContext::webPage):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::layerDidEnterContext):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::sendDidFirstLayerFlushIfNeeded):
(WebKit::TiledCoreAnimationDrawingArea::sendEnterAcceleratedCompositingModeIfNeeded):
(WebKit::TiledCoreAnimationDrawingArea::setLayerHostingMode):
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionManager::removeContext):
(WebKit::PlaybackSessionManager::setUpPlaybackControlsManager):
(WebKit::PlaybackSessionManager::contextIdForMediaElement):
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.h:
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::VideoFullscreenManager::setupRemoteLayerHosting):
(WebKit::VideoFullscreenManager::enterVideoFullscreenForVideoElement):
(WebKit::VideoFullscreenManager::setVideoLayerFrameFenced):

Co-authored-by: Arunsundar Kannan &lt;arunsundar_kannan@apple.com&gt;

Canonical link: <a href="https://commits.webkit.org/260575@main">https://commits.webkit.org/260575@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/89dd664d2c7bc5772b909242c59ebfb08fd1c0cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108392 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17486 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41254 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117507 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116871 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112278 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8766 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100664 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114158 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97420 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42183 "Passed tests") | 
| | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29065 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83883 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10316 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30409 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11061 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7315 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16465 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50005 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7312 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12647 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp [hunk 1]
```diff
@@ -1325,6 +1332,19 @@ void GraphicsLayerCA::setContentsToPlatformLayerHost(LayerHostingContextIdentifi
     noteLayerPropertyChanged(ContentsPlatformLayerChanged);
 }
 
+void GraphicsLayerCA::setContentsToVideoElement(HTMLVideoElement& videoElement, ContentsLayerPurpose purpose)
+{
+    auto hostingContextID = videoElement.layerHostingContextID();
+    if(!(hostingContextID == m_layerHostingContextID)) {
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp]: `if (auto hostingContextID = videoElement.layerHostingContextID != m_layerHostingContextID)` ?

File: Source/WebKit/Platform/cocoa/LayerHostingContext.h [hunk 2]
```diff
@@ -89,9 +91,15 @@ class LayerHostingContext {
...
     // with this context is desired.
     WTF::MachSendRight createFencePort();
+    
+    //Should be only be used inside webprocess
+    void updateCachedContextID(LayerHostingContextID contextID);
+    LayerHostingContextID cachedContextID();
 
 private:
     LayerHostingMode m_layerHostingMode;
+    //Denotes the contextID obtained from GPU process, should be returned for all calls to context ID in web process when UI side compositing is enabled. This is done to avoid making calls to CARenderServer from webprocess
```
Comment (eric-carlson) [inline Source/WebKit/Platform/cocoa/LayerHostingContext.h]: Nit: Space between `//` and `Denotes`. This is really long and should be wrapped.

File: Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm [hunk 3]
```diff
@@ -874,9 +946,19 @@ - (BOOL)prefersStatusBarHidden
 {
     auto& [model, interface] = ensureModelAndInterface(contextId);
 
-    [CATransaction flush];
+    auto videoView = model->layerHostView();
```
Comment (eric-carlson) [inline Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm]: Can this be moved inside of the `if (auto playerLayer = model->playerLayer())` block?

File: Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm [hunk 4]
```diff
@@ -219,6 +220,15 @@
             m_hostedLayers.remove(*hostIdentifier);
         }
     }
+
+    auto videoLayerIter = m_videoLayers.find(layerID);
+    if (videoLayerIter != m_videoLayers.end()) {
```
Comment (eric-carlson) [inline Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:228]: `if (auto videoLayerIter = m_videoLayers.find(layerID) != m_videoLayers())`

File: Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm [hunk 5]
```diff
@@ -117,11 +119,34 @@
...
+    m_livePlatformLayers.add(layerID, &layer);
+}
+
 void RemoteLayerTreeContext::layerWillLeaveContext(PlatformCALayerRemote& layer)
 {
     ASSERT(layer.layerID());
     GraphicsLayer::PlatformLayerID layerID = layer.layerID();
 
+    auto videoLayerIter = m_videoLayers.find(layerID);
+    if (videoLayerIter != m_videoLayers.end()) {
```
Comment (eric-carlson) [inline Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:147]: `if (auto videoLayerIter = m_videoLayers.find(layerID) != m_videoLayers())`

**Review by eric-carlson — APPROVED** [review]
r=me

File: Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp (hunk 1 above)
Comment (jernoble) [reply Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp]: I don't think you can initialize and check in the same if statement.

Comment (eric-carlson) [reply Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp]: It is done a lot in WebKit

Comment (eric-carlson) [reply Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp]: Actually you did it yourself in https://bugs.webkit.org/show_bug.cgi?id=162971

File: Source/WebCore/html/HTMLMediaElement.h [hunk 6]
```diff
@@ -61,6 +62,10 @@
 #include <wtf/StringPrintStream.h>
 #endif
 
+namespace WTF {
+class MachSendRight;
+}
+
```
Comment (pvollan) [inline Source/WebCore/html/HTMLMediaElement.h:68]: Is this declaration required? I can't seem to find a reference to it.

Comment (pvollan) [reply Source/WebCore/html/HTMLMediaElement.h:68]: Sorry, I do see it now :)

File: Source/WebKit/Shared/WebExtensionContextIdentifier.h [hunk 7]
```diff
@@ -25,8 +25,6 @@
 
 #pragma once
 
-#if ENABLE(WK_WEB_EXTENSIONS)
-
```
Comment (pvollan) [inline Source/WebKit/Shared/WebExtensionContextIdentifier.h:29]: This change looks somewhat unrelated, is it strictly needed?

File: Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp [hunk 8]
```diff
@@ -1490,6 +1493,15 @@ WTFLogChannel& MediaPlayerPrivateRemote::logChannel() const
 }
 #endif
 
+LayerHostingContextID MediaPlayerPrivateRemote::hostingContextID() const
+{
+    return m_layerHostingContextID;
+}
+void MediaPlayerPrivateRemote::setLayerHostingContextID(LayerHostingContextID inID)
```
Comment (pvollan) [inline Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp]: There's missing a new line between these methods.

File: Source/WebCore/platform/graphics/GraphicsLayer.h [hunk 9]
```diff
@@ -524,6 +535,7 @@ class GraphicsLayer : public RefCounted<GraphicsLayer> {
     virtual void setContentsToSolidColor(const Color&) { }
     virtual void setContentsToPlatformLayer(PlatformLayer*, ContentsLayerPurpose) { }
     virtual void setContentsToPlatformLayerHost(LayerHostingContextIdentifier) { }
+    virtual void setContentsToVideoElement(HTMLVideoElement&, ContentsLayerPurpose) { }
```
Comment (pvollan) [inline Source/WebCore/platform/graphics/GraphicsLayer.h:538]: In order to avoid referencing HTML element classes inside the graphics layer class, you could possibly create a new, more generic class to use here instead of HTMLVideoElement. This class would then declare the virtual methods needed, and HTMLVideElement could subclass and override the methods. However, I don't think this is needed to move forward with the current patch, and could be done in a follow-up patch, if you think it makes sense.

**Review by pvollan — APPROVED** [review]
R=me.

File: Source/WebCore/platform/graphics/GraphicsLayer.h [hunk 10]
```diff
@@ -33,6 +33,7 @@
 #include "FloatRoundedRect.h"
 #include "FloatSize.h"
 #include "GraphicsLayerClient.h"
+#include "HTMLMediaElementIdentifier.h"
```
Comment (smfr) [inline Source/WebCore/platform/graphics/GraphicsLayer.h:36]: This is a layering violation.

File: Source/WebCore/platform/graphics/GraphicsLayer.h (hunk 9 above)
Comment (smfr) [reply Source/WebCore/platform/graphics/GraphicsLayer.h:538]: GraphicsLayer cannot know about HMTL elements.

File: Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp [hunk 11]
```diff
@@ -1287,6 +1294,23 @@ void GraphicsLayerCA::setContentsToPlatformLayerHost(LayerHostingContextIdentifi
     noteLayerPropertyChanged(ContentsPlatformLayerChanged);
 }
 
+void GraphicsLayerCA::setContentsToVideoElement(HTMLVideoElement& videoElement, ContentsLayerPurpose purpose)
```
Comment (smfr) [inline Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:1297]: The way to fix the layering violation would be to call something on the HTMLVideoElement that sets up this GraphicsLayer.
